### PR TITLE
WIP: Stemplots

### PR DESCRIPTION
**Pull request by alexhallam**
Hi, I have a first rough draft, but it is a long way from production/merging. I have a lot more formatting to do and some tests to write. 

I have a couple of questions about formatting:

1. How do I add a solid line to separate the stem and the leaf? I am guessing it is a unicode character, but I am not sure. Currently I am using "|"

screen shot 2016-12-18 at 6 20 40 pm

2. Do we want expanded stemplots, collapsed stemplot, or allow user to choose. Example:
Given [10,11,30,31] should the plot be

Expanded:
```
1 | 01
2 |
3 | 01
```

Collapsed:
```
1 | 01
3 | 01
```

FWIW R seems to do collapsed, but I am fine with either way. It is currently running the expanded option.

Thanks for the guidance!

### CODE REVIEW

**Comment by Evizero** [comment]
Hi! Thanks for working on this.

Concerning the separating sign, `|` seems reasonable. You could make a keyword argument `symb = "|”`, which would allow a user to overwrite it.

Concerning collapsing, I think this would be a perfect situation for a keyword argument `collapse`, which could default to `true`

**Comment by alexhallam** [comment]
I will make those changes. Thanks for letting me contribute!

**Comment by alexhallam** [comment]
It has been I while since I pushed. Thought I would just give a little update. 

This is still not ready. I am working on making it more robust. I will have to introduce some rounding to handle decimals better.

Random Integers

2017-01-02-175419_891x458_scrot

Negative Integers

2017-01-02-175448_891x458_scrot

Floats :(

2017-01-02-175724_891x458_scrot

**Comment by Evizero** [comment]
Nice progress!

**Comment by alexhallam** [comment]
# Fixes

Plots are coming together quite nice. 

Commas have been removed from leaves, and negative signs now only show on the stems.

2017-01-12-152201_1275x458_scrot


Rounding was added. A scale was also introduced so the user can change the position of the decimal. 

2017-01-12-144829_891x458_scrot

# Bugs

1. -0 needs to be introduced when values are negative
2. Key needs to be accurate when the scale is changed

The below plot makes the -0 problem apparent.
 
2017-01-12-145025_1275x353_scrot

The below plot shows that the key is no longer useful when the scale is changed.

2017-01-12-145237_1275x879_scrot

# Final Thoughts

The key should not be a hard fix. The -0 problem has me stumped. I am trying to think of an elegant way to add it in without doing a major change to my current code. I hope to have something that can be merged by the end of the month.

**Comment by Evizero** [comment]
very nice. thanks for working on this

**Comment by Evizero** [comment]
To you use anything to compare against? `R` has `stem` which may be a good reference.

```R
> stem(sample(-100:100, 80))

  The decimal point is 1 digit(s) to the right of the |

   -8 | 976310742
   -6 | 84352
   -4 | 43176430
   -2 | 98214
   -0 | 97421642
    0 | 480126
    2 | 0124501347
    4 | 057123458
    6 | 0380123678
    8 | 034560246
   10 | 0
```

**Comment by alexhallam** [comment]
I have been comparing my examples against R exclusively. R has many more features that I would eventually like to add. 

My scale syntax is different. In R, when you want to change the scale you specify how long or wide you want the plot to be. At the moment my scale just has the user specify where the decimal goes. Would you like me to follow Rs syntax exactly? I don't mind. This is a fun side project for me.

**Comment by Evizero** [comment]
> Would you like me to follow Rs syntax exactly? I don't mind. This is a fun side project for me.

Nono, no need. I am open to changes if you think they make more sense. The only thing I'd say I like more from R is the indentation on the left hand side, which aligns more nicely.

**Comment by alexhallam** [comment]
The left indent is not a problem. 

Also, I will make sure that my outputs are a little more clean. R has very clean stem plots. I will make sure that`stemplot()` is kept to the same standard.

**Comment by alexhallam** [comment]
Just a couple more bugs to work out. I am getting close to wrapping this up. 

2017-01-24-060443_1083x458_scrot

**Comment by alexhallam** [comment]
This is Ready For Comments (RFC)

Here is the final product:

2017-01-24-072436_1083x929_scrot

The key correctly ascertains the position of the decimal. The description changes wording from "right" to "left" when appropriate. 

2017-01-24-072459_1083x929_scrot
 
Some fun with changing the divider.

2017-01-24-073327_1083x929_scrot

**Comment by Evizero** [comment]
please replace all tabs with 4 spaces

File: src/interface/stemplot.jl [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+A plot of object type 
+
+TODO
+
+Author(s)
+========
+- Iain Dunning (Github: https://github.com/IainNZ)
+- Christof Stocker (Github: https://github.com/Evizero)
+- Kenta Sato (Github: https://github.com/bicycle1885)
+- Alex Hallam (Github: https://github.com/alexhallam)
```
Comment (Evizero) [inline src/interface/stemplot.jl]: Did you reuse code from other places of the package? if not than you should be the only one listed here

**Comment by Evizero** [comment]
please add two or three plots at the end of `test/tst_plots.jl` just to make sure everything compiles and runs

**Comment by Evizero** [comment]
I still think a space before the divider (in the output) would be nice

**Comment by Evizero** [comment]
If you feel up to it, the cleanest way to integrate this plot type to the family would be in a similar way to how `barplot` does it, by having a `GraphicsArea` type that is then boxed into a `Plot` and decorated.

But this is not a requirement for merging. I am ok with the simpler approach that you did as well.

File: src/interface/stemplot.jl [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+`stemplot(rand(1:100,80))`
+
+`stemplot(rand(-100:100,300))`
+
+`stemplot(randn(50),scale = 1)`
+
+`stemplot(randn(50),scale = .1)`
+
+`stemplot(rand(1:59, 200),divider = ":")`
```
Comment (Evizero) [inline src/interface/stemplot.jl]: I'd prefer a single example, but with the output added. (example: https://github.com/Evizero/UnicodePlots.jl/blob/master/src/interface/barplot.jl#L59-L72)

**Comment by alexhallam** [comment]
1. tabs to 4 spaces - I probably change my vim settings to much. If this is still off just let me know.

2. edit authors - done

3. add three plots to `tst_plots.jl` - done

4. add space before divider - done

2017-01-24-150505_891x458_scrot


5. `GraphicsArea` - I will have to read more of your code to know how to implement this.

**Comment by Evizero** [comment]
nice. Thanks for doing this. I'll try it out later and if nothing obvious strikes my eye I'll merge. We can do the `GraphicsArea` change in a new PR if it is something your would be interested in doing, but I don't consider it blocking

**Comment by alexhallam** [comment]
Thanks for the help. I am very new to `Julia` so this was a fun project. I understand that this is a minimum viable product. If (when) any issues come up I am happy to continue maintaining this function. I may look into `GraphicsArea` - just not immediately. I agree that this would make the plots look better and more unified with the other plots in your package.

**Comment by Evizero** [comment]
There seems to be some test errors. This particular one is because you have to add `stemplot` to the list of exported functions here https://github.com/Evizero/UnicodePlots.jl/blob/master/src/UnicodePlots.jl#L41-L47

**Comment by alexhallam** [comment]
Sorry, I am a little confused at what is happening here. I can run the test just fine after the change you suggested on my personal computer, but Travis CI does not seem to see the stemplot.jl file still.

**Comment by Evizero** [comment]
Seems to be related to changes in 0.6. We can live with them for now and fix later

**Comment by Evizero** [comment]
There seems to be some unexpected behaviour when comparing to R output. is this on purpose?

```julia
julia> stemplot(rand(100)*10)

    0 | 0000000000000111111111111112222222222223333333333334444444444555555666666666677778888888899999999999

  Key: 0|0 = 0.0
  Description: The decimal is 1 digit(s) to the right of |

julia> stemplot(rand(100)*10, scale=1)

    0 | 00000000000
    1 | 00000000000000
    2 | 00000000
    3 | 000000
    4 | 00000000
    5 | 0000000000000000000
    6 | 00000000
    7 | 00000000
    8 | 0000000000000
    9 | 00000

  Key: 9|0 = 9.0
  Description: The decimal is 0 digit(s) to the right of |
```

R output:

```R
> stem(runif(100)*10)

  The decimal point is at the |

  0 | 1122456668
  1 | 01112346778899
  2 | 56999
  3 | 01337899
  4 | 35566789
  5 | 0112233456777799
  6 | 2234679
  7 | 01112446667
  8 | 11233456667799
  9 | 0046779
```

**Comment by Evizero** [comment]
Also, looking at the document locally I see a lot of trailing white spaces. See if you can configure your editor to strip trailing white spaces.

**Comment by alexhallam** [comment]
I just added a quick fix. I am sure there are more problems. I will keep playing with it. When you see something else that is odd let me know, so I can add it to the list of things to test. Thanks!

File: test/tst_plots.jl [hunk 3]
```diff
@@ -240,6 +240,8 @@ miny = -1.2796649117521434e218
 maxy = -miny
 println(scatterplot([1],[miny],xlim=[1,1],ylim=[miny,maxy],title="Don't you crash on me!"))
 
+include("../src/interface/stemplot.jl")
```
Comment (Evizero) [inline test/tst_plots.jl]: that is never a good idea, because we don't want a user to require having to do the same thing

**Comment by Evizero** [comment]
Remind me. Where were we with this PR? Are you waiting for me to perform a review/merge? Sorry I lost track a little

**Comment by alexhallam** [comment]
Some bugs were introduced with Julia 6.0. Go ahead and close this PR. I will work on stemplots as an independent repo that will be compatible with 6.0. When things are working well, I will have you take a look at it (probably via a new PR) and determine of it is something you would still like to see in UnicodePlots.

**Pull request closed without merging**